### ISSUE TITLE
VM virtual disk files - not deduping 

### ISSUE THREAD

**Issue by luisBSBDATA** [duplicati/duplicati#3255]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.3.2
- **Operating system**: Debian
- **Backend**: WebDAV and FTPS

## Description

When backing up VM virtual disks (vmdk from Proxmox and img from QNAP Virtualization Station) Duplicati always copy the full file just like if it was all modified data, instead of detecting the old data already copied and copying only the new data. This results on the wonderful Duplicati deduplication being useless for this cases.

## Steps to reproduce

1. Set up the task
2. Execute the task
3. Data is being treated as "Modified data" instead of "Old data"

- **Actual result**:
Duplicati copies all the VM virtual disk file everytime instead of deduping
- **Expected result**:
Duplicati deduping the VM virtual disk file and only copying the new data

## Screenshots

Not necessary (If you think it's relevant i'll post whatever you guys need)

## Debug log

Not necessary (If you think it's relevant i'll post whatever you guys need)

**Comment by Pectojin** [comment]
It likely is deduplicating the vm disk, but it's just doing a poor job.

Duplicati's deduplicati is quite simple. It takes whatever file you want backed up and splits it into pieces of `block size`. It does this from start to end without considering what the data looks like. The problem is that if you add an extra bit in the beginning of the file then every piece is shifted one bit and no longer match the pieces from the last backup.

In practice this means you get most of your files deduplicated even if some of it is re-uploaded. However with VM's it's often the case that the vm disk file is very large with frequent small things all over the disk. This causes large parts of the file to be reuploaded. 

You would see much the same behavior if you back up a 50GB text file and keep prepending a letter in the beginning of the file between each backup.

The solution is to rework the deduplication algorithm or to add special deduplication for vm disks (since it's often only a problem for these). However, I haven't heard of any work being done on this yet.

**Comment by luisBSBDATA** [comment]
Yeah I readed that somewhere, thanks for clarifying it. Anyways would be nice to implement something for this kind of files, VM disks are normally big files so deduplication would be a great thing

**Comment by zertrin** [comment]
Just leaving this this here, a quite interesting explanation about rolling hashes with links at the bottom for further reading: https://joelearnsopensource.wordpress.com/2016/04/13/deduplication-part-1-rabin-karp-for-variable-chunking/

**Comment by luisBSBDATA** [comment]
Nice info, I hope it's useful for the devs.

**Comment by benkuly** [comment]
Does this problem still appears in the current versions of duplicati?

**Comment by ts678** [comment]
> Does this problem still appears in the current versions of duplicati?

The chunking hasn't changed, however I'd like to hear some actual technical detail that early "insert" problem is the problem. This would require someone looking through the mentioned image formats "vmdk from Proxmox and img from QNAP Virtualization Station". In my opinion it would be unusual for an image on disk to have that happen to it routinely, because that would be ridiculously inefficient.

Ultimately the image must be kept in a file, and there's no file operation AFAIK called "insert here"... Looking further down into the sectors on storage, you'd have to rewrite everything to do that insert.

I would much more agree with "changes all over" ideas, and one change anywhere may change the timestamp on the huge file, and require Duplicati to read through the whole file to find the change. Really sophisticated programs such as Veeam apparently do "Changed Block Tracking" with a driver. Taking a general-purpose approach, Duplicati would have to read through the whole file if nothing changed except you changing its timestamp however your OS does that e.g. Linux touch command. 

Looking at your backup report at "SizeOfOpenedFiles" versus "BytesUploaded" may be informative.

Big databases may have similar issues of changes all over and having to figure out where they are... 
One thing that some databases do that variable chunking "might" help is to dump to a backup file, rather than have the backup program take a (probably not application-consistent) block snapshot. 
With such an approach, a single small change might be steered around with variable chunking, but whether or not variable chunking could steer around more and big change stays an open question.
